### PR TITLE
feat: add mtls support for NetHttpTransport

### PR DESCRIPTION
**Pull request by arithmetic1728**
go/java-apiary-client-mtls (section 2.2.2.1)

Support 
(1) passing keystore to SslContext constructor for client certificate and private key.
(2) creating keystore from certAndKey input stream.

Summary of new functions:
```
// In google-http-client/src/main/java/com/google/api/client/util/SecurityUtils.java
// create mtls key store from client cert and key input stream
public static KeyStore createMtlsKeyStore(InputStream certAndKey)

// In google-http-client/src/main/java/com/google/api/client/util/SslUtils.java
// init sslcontext with trust store and mtls key store
public static SSLContext initSslContext(
      SSLContext sslContext,
      KeyStore trustStore,
      TrustManagerFactory trustManagerFactory,
      KeyStore mtlsKeyStore,
      String mtlsKeystorePassword,
      KeyManagerFactory keyManagerFactory)

// In google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java
public Builder trustCertificates(
        KeyStore trustStore, KeyStore mtlsKeyStore, String mtlsKeystorePassword)

// Indict if transport is mtls.
public boolean isMtls() {
    return this.isMtls;
}
``

### CODE REVIEW

File: google-http-client/src/main/java/com/google/api/client/http/HttpTransport.java
Comment (chingor13) [inline google-http-client/src/main/java/com/google/api/client/http/HttpTransport.java:139]: I don't see anywhere this is being used, or set to anything other than `false`. Do we actually need to expose this?

File: google-http-client/src/main/java/com/google/api/client/util/SslUtils.java
Comment (chingor13) [inline google-http-client/src/main/java/com/google/api/client/util/SslUtils.java:132]: `char[]` is not a friendly Java construct. Consider `String` and call `toCharArray()`

**Review by chingor13 — CHANGES_REQUESTED** [review]
We also need to add tests for this

File: google-http-client/src/main/java/com/google/api/client/http/HttpTransport.java
Comment (arithmetic1728) [reply google-http-client/src/main/java/com/google/api/client/http/HttpTransport.java:139]: it will be used in the generated client code for api endpoint auto switch (auto switch to mtls endpoint if transport is mtls)

File: google-http-client/src/main/java/com/google/api/client/util/SslUtils.java
Comment (arithmetic1728) [reply google-http-client/src/main/java/com/google/api/client/util/SslUtils.java]: done

**Comment by arithmetic1728** [comment]
> We also need to add tests for this

Okay, will add tests.

**Comment by arithmetic1728** [comment]
> We also need to add tests for this

Tests are added now.

File: google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpTransportTest.java
Comment (chingor13) [inline google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpTransportTest.java:56]: These 2 tests should be split into separate tests and the test name should indicate what condition it's trying to test.

File: google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java
Comment (chingor13) [inline google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java:305]: Builder setters should be `setSomething()` or possibly `applySomething()`. The documentation should also indicate that this affects MTLS configuration.

**Review by chingor13 — CHANGES_REQUESTED** [review]
We're going to hold off on approving and merging this approach (especially the `isMtls()` accessor) until the whole design is complete and approved.

File: google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpTransportTest.java
Comment (arithmetic1728) [reply google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpTransportTest.java]: done

File: google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java
Comment (arithmetic1728) [reply google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java:305]: This file has a couple of `trustCertificates` like methods to set SSL socket. I am using the same name to be consistent with those methods.

Documentation is updated.

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (chingor13) [inline google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java:221]: For this test, maybe add the combined cert/private key file that we expect to parse rather than having 2 files then concatenating them in code and creating a new ByteArrayInputStream.

**Review by chingor13 — CHANGES_REQUESTED** [review]
Looking really close.

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (arithmetic1728) [reply google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java]: Done.

**Review by elharo — COMMENTED** [review]
For open source projects like this one, it's very helpful to have designs and issues on Github rather than in corp.

**Comment by chanseokoh** [comment]
Looks like this is to enable mutual TLS for `NetHttpTransport`. How about #904 (`ApacheHttpTransport`)? Is it that #904 should eventually be fixed?

**Comment by chingor13** [comment]
MTLS support will also need to be implemented for the Apache transport as well.

**Comment by arithmetic1728** [comment]
> For open source projects like this one, it's very helpful to have designs and issues on Github rather than in corp.

I added a summary of the new functions added, I think they are self explanatory. The design doc is for client libs, http client is just a building block.

**Comment by arithmetic1728** [comment]
> Looks like this is to enable mutual TLS for `NetHttpTransport`. How about #904 (`ApacheHttpTransport`)? Is it that #904 should eventually be fixed?

yes, mTLS support will be added to apache transport.

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (chingor13) [inline google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java:214]: In `assertEquals`, the expected value comes first.

```suggestion
    assertEquals(1, mtlsKeyStore.size());
```

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (chingor13) [inline google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java:183]: Use a try/catch block for asserting exceptions thrown:

```java
boolean thrown = false;
try {
  SecurityUtils.createMtlsKeyStore(certMissing);
  fail("should have thrown");
} catch (IllegalArgumentException e) {
  assertTrue(e.getMessage().contains("some message"));
  thrown = true;
}
assertTrue(thrown, "should have caught an IllegalArgumentException");
```

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (chingor13) [inline google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java:200]: Try/catch

**Review by chingor13 — CHANGES_REQUESTED** [review]
A few nits

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (arithmetic1728) [reply google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java]: done

File: google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java
Comment (arithmetic1728) [reply google-http-client/src/test/java/com/google/api/client/util/SecurityUtilsTest.java]: done

File: google-http-client/src/main/java/com/google/api/client/util/SecurityUtils.java
Comment (arithmetic1728) [reply google-http-client/src/main/java/com/google/api/client/util/SecurityUtils.java:268]: modify the code, only set certSection if it is null, so the first cert is used.

File: google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java
Comment (arithmetic1728) [reply google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpTransport.java:300]: `createMtlsKeyStore` is a new function to load mtls cert and key. `loadKeyStore` is used to load CA cert. 

Modified the doc to add an example.
```
@param mtlsKeyStore key store for client certificate and key to establish mutual TLS. (use
           for example {@link SecurityUtils#createMtlsKeyStore(InputStream)})
```

**Review by chingor13 — APPROVED** [review]

**Comment by chingor13** [comment]
@arithmetic1728 Let's follow up with a PR to add `@Beta` annotations to the new APIs and `@since 1.38` to the javadoc of the new APIs

**Pull request merged**
